x-pack/plugins/security/server/authentication/providers/http.ts

```diff
@@ -79,6 +95,21 @@ export class HTTPAuthenticationProvider extends BaseAuthenticationProvider {
       this.logger.debug(
         `Request to ${request.url.pathname}${request.url.search} has been authenticated via authorization header with "${authorizationHeader.scheme}" scheme.`
       );
+
+      // If Kibana is configured to restrict JWT authentication only to selected routes, ensure that the route is marked
+      // with the `ROUTE_TAG_ACCEPT_JWT` tag to bypass that restriction.
+      if (
+        user.authentication_realm.type === JWT_REALM_TYPE &&
+        this.jwt?.taggedRoutesOnly &&
+        !request.route.options.tags.includes(ROUTE_TAG_ACCEPT_JWT)
+      ) {
+        this.logger.error(
+          `Attempted to authenticate with JWT credentials against ${request.url.pathname}${request.url.search}, but it's not allowed. ` +
+            `Ensure that the route is defined with the "${ROUTE_TAG_ACCEPT_JWT}" tag.`
+        );
+        return AuthenticationResult.notHandled();
+      }
```

Comment:
I'm torn on the location & timing of this check. This will allow a user to attempt JWT auth for disallowed routes, incurring unnecessary overhead (which may or may not matter). It also gives us different behavior in the failure scenario where the user presents an invalid JWT (`catch` block below), as we are relying on the successful auth attempt to detect if it should be allowed or not.

Performing this check earlier in the lifecycle would require us to inspect the shape of the `Bearer` token that we receive, to see if it looks like a JWT or not. That feels more fragile to me, but it eliminates the round-trip to ES. The fragility might be tolerable since this is scoped only to serverless, where we have full control over the types of values we accept in the `Authorization` header; In serverless, it'll only ever be one of [API Key, ES Token, JWT], whereas on-prem may be configured with a custom realm that we aren't equipped to detect.

I don't have a definitive opinion at this point, but I wanted to start a discussion to see how you/the team felt.

Reply:
> This will allow a user to attempt JWT auth for disallowed routes, incurring unnecessary overhead (which may or may not matter)

I'm not particularly concerned about this aspect. If someone truly wishes to introduce overhead to the cluster, they can already achieve that by using a custom `SAMLResponse` XML document, which could be large and/or expensive to decode. Unfortunately, there's not much we can do to mitigate this.

In fact, I can even see the benefit. This way, we'll catch all (expected and unexpected) authentication attempts in the Elasticsearch audit logs, together with all other authentication related audit logs. It might help us to spot this suspicious activity early on. What do you think?

> It also gives us different behavior in the failure scenario where the user presents an invalid JWT (catch block below), as we are relying on the successful auth attempt to detect if it should be allowed or not.

I believe we'll only have different logs, the user will get the same `401 Unauthorized` in both cases, or do you mean something else?

Reply:
> In fact, I can even see the benefit. This way, we'll catch all (expected and unexpected) authentication attempts in the Elasticsearch audit logs, together with all other authentication related audit logs. It might help us to spot this suspicious activity early on. What do you think?

I think you've convinced me. Sounds good!

> I believe we'll only have different logs, the user will get the same 401 Unauthorized in both cases, or do you mean something else?

Correct, just different logs

Reply:
> I think you've convinced me. Sounds good!

Great, thanks!